pkg/cloud/aws/actuators/machine/actuator.go

```diff
@@ -167,20 +119,84 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 	}
 
 	controlPlaneMachines := GetControlPlaneMachines(clusterMachines)
+	if len(controlPlaneMachines) == 0 {
+		return errors.New("unable to proceed with machine creation until there is at least 1 control plane machine")
+	}
 
-	isNodeJoin, err := a.isNodeJoin(scope, controlPlaneMachines)
-	if err != nil {
-		return errors.Wrapf(err, "failed to determine whether machine %q should join cluster %q", machine.Name, cluster.Name)
+	/*
+	If new machine is a control plane machine, we may be in the following states:
+
+	1) We have not yet attempted to create any control plane VMs
+	2) We have created our first control plane VM, but it's not ready yet (apiserver still booting)
+	3) We have created at least one control plane VM and it's ready
+
+	To avoid race conditions, we use a single ConfigMap to coordinate.
+
+	1) Absence of ConfigMap: if the current machine is a control plane machine, try to create the ConfigMap with that machine's name in the
+	   data.
+			If creation succeeds, proceed to create the VM for the chosen machine.
+			If creation fails because it already exists, move on to checking if the control plane is ready.
+
+	2) Presence of ConfigMap: the chosen control plane machine should either be booting or ready. Try to reach the apiserver.
+			If we can reach the apiserver, we can proceed with a join (can be a control plane or regular node)
+			If we can't reach the apiserver, we need to keep waiting.
+	 */
+
+
+	// Except for the very first control plane machine, everything will be a join, so let's assume we're joining.
+	isNodeJoin := true
+
+	// TODO make sure this name isn't longer than the max allowed length
+	controlPlaneSyncConfigMapName := fmt.Sprintf("%s-controlplane", cluster.Name)
+
+	controlPlaneConfigMap, err := a.coreClient.ConfigMaps(cluster.Namespace).Get(controlPlaneSyncConfigMapName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		if machine.Spec.Versions.ControlPlane == "" {
```

Comment:
Is this the canonical way to tell if it's a control plane machine? I thought CAP* relied on `ObjectMeta.Labels["set"]` to distinguish.

Reply:
Either is probably correct. The implementation in CAPA for `GetControlPlaneMachines` uses this method.

Reply:
So had a long discussion about this with @detiber a while back - https://kubernetes.slack.com/archives/C8TSNPY4T/p1560807049390100.

Basically `set` isn't part of CAPI, and there's nowhere that CAPA *uses* `set` that couldn't be the formal way of checking via the version string. In fact, there's a CAPI function, [`IsControlPlane`](https://github.com/kubernetes-sigs/cluster-api/tree/master/pkg/util/util.go#L134-L137), that does just that.

In other words, wherever CAPA uses `set`, it should likely switch to checking the Versions information.

Reply:
I prefer the explicit check against versions (at least for v1alpha1). If we do standardize on labels for v1alpha2+, then I would prefer we choose something namespaced and more descriptive for the key than `set`.

Reply:
For the sake of consistency in CAPA, I'd use the `set` label here to check if the machine is a control plane. Later, when we move to v1alpha2 standards to determine the role of a Machine, we can quickly find and replace.

Reply:
I think @vincepri might win here - we call `machine.Role()` to determine if it's a `"controlplane"` machine or not, and the implementation of `Role()` just grabs the `set` label.

I'll add a TODO

Reply:
Actually, probably don't need the TODO since this logic ultimately will move out of CAPA for v1alpha2.

Reply:
CAPA uses both `set` and the version in different places, and the CAPI function uses the version. Just saying :)

Reply:
True, but I do think the same logic that determines if it's a control plane init/join should be used here (which is checking `set`).